src/core/common/heap_array.hpp

```diff
@@ -0,0 +1,529 @@
+     *
+     * The allocated buffer and array capacity are automatically increased (by the `Array` itself) when new elements
+     * are added to array. Removing elements does not change the buffer and the capacity. A desired capacity can be
+     * reserved using `ReserveCapacity()` method.
+     *
+     * @returns The current capacity of the array.
+     *
+     */
+    IndexType GetCapacity(void) const { return mCapacity; }
+
+    /**
+     * This method allocates buffer to reserve a given capacity for array.
+     *
+     * If the requested @p aCapacity is smaller than the current length of the array, capacity remains unchanged.
+     *
+     * @param[in] aCapacity   The target capacity for the array.
+     *
+     * @retval kErrorNone     Array was successfully updated to support @p aCapacity.
+     * @retval kErrorNoBufs   Could not allocate buffer.
+     *
+     */
+    Error ReserveCapacity(IndexType aCapacity) { return Allocate(aCapacity); }
+
+    /**
+     * This method sets the array by taking the buffer from another given array (using move semantics).
+     *
+     * @param[in] aOther    The other `Heap::Array` to set from (rvalue reference).
+     *
+     */
+    void SetFrom(Array &&aOther)
+    {
+        Free();
+        mArray           = aOther.mArray;
+        mLength          = aOther.mLength;
+        mCapacity        = aOther.mCapacity;
+        aOther.mArray    = nullptr;
+        aOther.mLength   = 0;
+        aOther.mCapacity = 0;
+    }
+
+    /**
+     * This method overloads the `[]` operator to get the element at a given index.
+     *
+     * This method does not perform index bounds checking. Behavior is undefined if @p aIndex is not valid.
+     *
+     * @param[in] aIndex  The index to get.
+     *
+     * @returns A reference to the element in array at @p aIndex.
+     *
+     */
+    Type &operator[](IndexType aIndex) { return mArray[aIndex]; }
+
+    /**
+     * This method overloads the `[]` operator to get the element at a given index.
+     *
+     * This method does not perform index bounds checking. Behavior is undefined if @p aIndex is not valid.
+     *
+     * @param[in] aIndex  The index to get.
+     *
+     * @returns A reference to the element in array at @p aIndex.
+     *
+     */
+    const Type &operator[](IndexType aIndex) const { return mArray[aIndex]; }
+
+    /**
+     * This method gets a pointer to the element at a given index.
+     *
+     * Unlike `operator[]`, this method checks @p aIndex to be valid and within the current length. The returned
+     * pointer is valid only while the `Array` remains unchanged.
+     *
+     * @param[in] aIndex  The index to get.
+     *
+     * @returns A pointer to element in array at @p aIndex or `nullptr` if @p aIndex is not valid.
+     *
+     */
+    Type *At(IndexType aIndex) { return (aIndex < mLength) ? &mArray[aIndex] : nullptr; }
```

Comment:
Do we really want to tolerate the `out of range access` error case? I think we still require the caller to explicitly check that `aIndex` is in the range before calling this method, so maybe we should just assert here.

From another point, I think it's wrong to return `nullptr` for out of range access. For example, if the `Type` is a pointer and `mArray[aIndex]` is a `nullptr`, the user will not be able to tell it with the `out of range access` case. Thoughts?

Reply:
> From another point, I think it's wrong to return nullptr for out of range access. For example, if the Type is a pointer and mArray[aIndex] is a nullptr, the user will not be able to tell it with the out of range access case. Thoughts?

Just realized that `Type` cannot be a pointer with current implementation.

Reply:
We basically provide two ways to access items in the array
- We have `operator[]` which does not do any bound checks (up to caller to validate) and undefined behavior if index is out of bounds.
- `At()` which does bound checking (safer to use).

I tried to mirror these to be similar to what we have in `ot::Array()` class (which is fixed length array) and it is also similar to how `std::array` API works (`at()` does bound checking and `operator[]` does not).

In general, I prefer to avoid adding asserts in the simpler utility types (to keep them more flexible and smaller code size) and leave it to caller/user of the type to decide if they want to add checks.

-------

Regarding `Type` being pointer itself, I think it can work fine.

If `Type` itself is pointer then `At()` returns a "pointer to a pointer", and caller can tell the difference between pointer being `nullptr` (entry not found in the array) or `*ptr` being `nullptr` (the entry is in array but its value itself is `nullptr`).

Reply:
> If Type itself is pointer then At() returns a "pointer to a pointer", and caller can tell the difference between pointer being nullptr (entry not found in the array) or *ptr being nullptr (the entry is in array but its value itself is nullptr).

@abtink , when `Type` is a pointer and `aArray->At(5)` returns `nullptr`, can the caller tell whether the length of the array is smaller than 5 or value of `aArray[5]` is actually `nullptr`?

Reply:
@wgtdkp , yes the caller can distinguish this. Note that `At()` returns a  pointer to `Type` (so we have an indirection) and not the `Type` itself.

For example if `Type` is `char *`, then `At()` is `char **At(IndextType)` which returns `char **` (not `char *`).
So for `char **ptr = At(5)`:
- If `(ptr == nullptr)` then caller knows there was no entry at index 5.
- If `(ptr != nullptr)`, then `*ptr` gives the value of `char *` stored in the array and if we have `(*ptr == nullptr)`, then it indicates our `Array<char *>` is holding a `nullptr` at its 5th index.